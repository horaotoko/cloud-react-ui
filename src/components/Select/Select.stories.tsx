import { useState } from "react";
import { Select as SelectComponent, Option, GroupOption } from "./Select";
import { AsyncSelect as AsyncSelectComponent } from "./AsyncSelect";
import { boolean, text } from "@storybook/addon-knobs";

export default {
  component: SelectComponent,
  title: "components/Select",
};

const defaultOptions: Option[] = [
  {
    value: "1",
    label: "label1",
  },
  {
    value: "2",
    label: "label2",
  },
  {
    value: "3",
    label: "label3",
  },
  {
    value: "4",
    label: "label4",
  },
];

const groupOptions: GroupOption[] = [
  {
    label: "parent1",
    options: [
      {
        label: "child1",
        value: "child1",
      },
      {
        label: "child2",
        value: "child2",
      },
    ],
  },
  {
    label: "parent2",
    options: [
      {
        label: "child3",
        value: "child3",
      },
      {
        label: "child4",
        value: "child4",
      },
    ],
  },
];

const Indicator = (
  <svg
    width="12"
    height="12"
    viewBox="0 0 12 12"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      fillRule="evenodd"
      clipRule="evenodd"
      d="M8.33333 7.33333H7.80667L7.62 7.15333C8.27333 6.39333 8.66667 5.40667 8.66667 4.33333C8.66667 1.94 6.72667 0 4.33333 0C1.94 0 0 1.94 0 4.33333C0 6.72667 1.94 8.66667 4.33333 8.66667C5.40667 8.66667 6.39333 8.27333 7.15333 7.62L7.33333 7.80667V8.33333L10.6667 11.66L11.66 10.6667L8.33333 7.33333ZM4.33333 7.33333C2.67333 7.33333 1.33333 5.99333 1.33333 4.33333C1.33333 2.67333 2.67333 1.33333 4.33333 1.33333C5.99333 1.33333 7.33333 2.67333 7.33333 4.33333C7.33333 5.99333 5.99333 7.33333 4.33333 7.33333Z"
      fill="#7C8291"
    />
  </svg>
);

export const Select = (): JSX.Element => {
  const [selected, setSelected] = useState<Option>();
  const isGroupOption = boolean("GroupOption", false);

  return (
    <SelectComponent
      instanceId="select"
      options={isGroupOption ? groupOptions : defaultOptions}
      value={selected}
      defaultValue={
        isGroupOption ? groupOptions[0].options[0] : defaultOptions[0]
      }
      placeholder={text("Placeholder", "")}
      onChange={(option) => setSelected(option)}
      clearable={boolean("Clearable", true)}
      disabled={boolean("Disabled", false)}
      searchable={boolean("Searchable", true)}
      error={boolean("Error", false)}
      indicatorImage={boolean("customIndicator", false) ? Indicator : undefined}
    />
  );
};

export const AsyncSelect = (): JSX.Element => {
  const [selected, setSelected] = useState<Option>();

  const handleInputChange = async (inputValue: string) => {
    return new Promise<Option[]>((resolve) => {
      setTimeout(
        () =>
          resolve(
            defaultOptions.filter((option) =>
              option.label.toLowerCase().includes(inputValue.toLowerCase())
            )
          ),
        1000
      );
    });
  };

  return (
    <AsyncSelectComponent
      defaultOptions
      value={selected}
      onChange={(option) => setSelected(option)}
      onInputChange={handleInputChange}
    />
  );
};
