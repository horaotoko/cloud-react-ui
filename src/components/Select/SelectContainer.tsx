import { SelectPresenter } from "./SelectPresenter";

export type GroupOption = {
  label: string;
  options: Option[];
};

export type Option = {
  value: string;
  label: string;
  isDisabled?: boolean;
};

export type SelectProps = {
  options: Option[] | GroupOption[];
  value?: Option | GroupOption;
  name?: string;
  placeholder?: string;
  onChange?: (option: Option) => void;
  clearable?: boolean;
  disabled?: boolean;
  searchable?: boolean;
  error?: boolean;
  className?: string;
};

export function SelectContainer(props: SelectProps): React.ReactElement {
  return <SelectPresenter {...props} />;
}
