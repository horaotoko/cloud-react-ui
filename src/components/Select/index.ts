export { Select } from "./Select";
export type { Props as SelectProps } from "./Select";

export { AsyncSelect } from "./AsyncSelect";
export type { Props as AsyncSelectProps } from "./AsyncSelect";
