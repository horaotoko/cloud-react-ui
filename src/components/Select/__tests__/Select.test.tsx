import { render } from "@testing-library/react";
import { Select } from "../Select";

describe("Select", () => {
  it("default", () => {
    const { asFragment, container } = render(<Select />);
    expect(asFragment()).toMatchSnapshot();

    expect(
      container.getElementsByClassName("react-select__control")[0]
    ).toHaveStyle("background-color: #FFFFFF");
  });

  it("error props", () => {
    const { container } = render(<Select error={true} />);

    expect(
      container.getElementsByClassName("react-select__control")[0]
    ).toHaveStyle("background-color: #FFF4F4");
  });
});
