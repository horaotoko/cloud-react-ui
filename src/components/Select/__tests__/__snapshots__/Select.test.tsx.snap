// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select default 1`] = `
<DocumentFragment>
  .c0 {
  display: inline-block;
  width: 100%;
}

.c0 .react-select__control {
  width: 100%;
  height: 34px;
  background-color: #FFFFFF;
  border-radius: 4px;
  border-width: 1px;
  border-style: solid;
  border-color: #CCC;
  color: #333333;
  font-size: 13px;
  cursor: default;
  box-shadow: none;
}

.c0 .react-select__control:hover {
  box-shadow: 0 1px 0 rgba(0,0,0,0.06);
  border-color: #CCC;
}

.c0 .react-select__control--is-disabled {
  cursor: not-allowed;
  pointer-events: auto;
  background-color: #EEEEEE;
}

.c0 .react-select__placeholder {
  color: #999999;
}

.c0 .react-select__value-container {
  height: 100%;
}

.c0 .react-select__menu {
  width: calc(100% - 2px);
  top: calc(100% - 3px);
  left: 1px;
  margin: 0;
  border-radius: 0 0 4px 4px;
  font-size: 13px;
  z-index: 101;
}

.c0 .react-select__menu-list {
  padding: 0;
}

.c0 .react-select__option--is-focused {
  color: #333333;
  background-color: #EDF5FE;
}

.c0 .react-select__option--is-selected {
  color: #333333;
  background-color: #F5FAFF;
}

.c0 .react-select__option--is-disabled {
  color: #CCC;
}

.c0 .react-select__option:last-child {
  border-radius: 0 0 4px 4px;
}

.c0 .react-select__indicator-separator {
  display: none;
}

.c0 .react-select__clear-indicator {
  padding-right: 0;
}

.c0 .react-select__clear-indicator svg {
  width: 17px;
  height: 17px;
}

.c0 .react-select__clear-indicator:hover {
  color: #D0021B;
}

.c0 .react-select__dropdown-indicator {
  padding-left: 3px;
}

.c0 .react-select__dropdown-indicator svg {
  width: 16px;
  height: 16px;
}

<div
    class="c0 css-2b097c-container"
  >
    <div
      class="react-select__control css-yk16xz-control"
    >
      <div
        class="react-select__value-container css-g1d714-ValueContainer"
      >
        <div
          class="react-select__placeholder css-1wa3eu0-placeholder"
        >
          Select...
        </div>
        <div
          class="css-b8ldur-Input"
        >
          <div
            class="react-select__input"
            style="display: inline-block;"
          >
            <input
              aria-autocomplete="list"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              id="react-select-2-input"
              spellcheck="false"
              style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
              tabindex="0"
              type="text"
              value=""
            />
            <div
              style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
            />
          </div>
        </div>
      </div>
      <div
        class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <span
          class="react-select__indicator-separator css-1okebmr-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Select with indicatorImage props 1`] = `
<DocumentFragment>
  .c0 {
  display: inline-block;
  width: 100%;
}

.c0 .react-select__control {
  width: 100%;
  height: 34px;
  background-color: #FFFFFF;
  border-radius: 4px;
  border-width: 1px;
  border-style: solid;
  border-color: #CCC;
  color: #333333;
  font-size: 13px;
  cursor: default;
  box-shadow: none;
}

.c0 .react-select__control:hover {
  box-shadow: 0 1px 0 rgba(0,0,0,0.06);
  border-color: #CCC;
}

.c0 .react-select__control--is-disabled {
  cursor: not-allowed;
  pointer-events: auto;
  background-color: #EEEEEE;
}

.c0 .react-select__placeholder {
  color: #999999;
}

.c0 .react-select__value-container {
  height: 100%;
}

.c0 .react-select__menu {
  width: calc(100% - 2px);
  top: calc(100% - 3px);
  left: 1px;
  margin: 0;
  border-radius: 0 0 4px 4px;
  font-size: 13px;
  z-index: 101;
}

.c0 .react-select__menu-list {
  padding: 0;
}

.c0 .react-select__option--is-focused {
  color: #333333;
  background-color: #EDF5FE;
}

.c0 .react-select__option--is-selected {
  color: #333333;
  background-color: #F5FAFF;
}

.c0 .react-select__option--is-disabled {
  color: #CCC;
}

.c0 .react-select__option:last-child {
  border-radius: 0 0 4px 4px;
}

.c0 .react-select__indicator-separator {
  display: none;
}

.c0 .react-select__clear-indicator {
  padding-right: 0;
}

.c0 .react-select__clear-indicator svg {
  width: 17px;
  height: 17px;
}

.c0 .react-select__clear-indicator:hover {
  color: #D0021B;
}

.c0 .react-select__dropdown-indicator {
  padding-left: 3px;
}

.c0 .react-select__dropdown-indicator svg {
  width: 16px;
  height: 16px;
}

<div
    class="c0 css-2b097c-container"
  >
    <div
      class="react-select__control css-yk16xz-control"
    >
      <div
        class="react-select__value-container css-g1d714-ValueContainer"
      >
        <div
          class="react-select__placeholder css-1wa3eu0-placeholder"
        >
          Select...
        </div>
        <div
          class="css-b8ldur-Input"
        >
          <div
            class="react-select__input"
            style="display: inline-block;"
          >
            <input
              aria-autocomplete="list"
              autocapitalize="none"
              autocomplete="off"
              autocorrect="off"
              id="react-select-3-input"
              spellcheck="false"
              style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
              tabindex="0"
              type="text"
              value=""
            />
            <div
              style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
            />
          </div>
        </div>
      </div>
      <div
        class="react-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <span
          class="react-select__indicator-separator css-1okebmr-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer"
        >
          <svg
            fill="none"
            height="12"
            viewBox="0 0 12 12"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8.33333 7.33333H7.80667L7.62 7.15333C8.27333 6.39333 8.66667 5.40667 8.66667 4.33333C8.66667 1.94 6.72667 0 4.33333 0C1.94 0 0 1.94 0 4.33333C0 6.72667 1.94 8.66667 4.33333 8.66667C5.40667 8.66667 6.39333 8.27333 7.15333 7.62L7.33333 7.80667V8.33333L10.6667 11.66L11.66 10.6667L8.33333 7.33333ZM4.33333 7.33333C2.67333 7.33333 1.33333 5.99333 1.33333 4.33333C1.33333 2.67333 2.67333 1.33333 4.33333 1.33333C5.99333 1.33333 7.33333 2.67333 7.33333 4.33333C7.33333 5.99333 5.99333 7.33333 4.33333 7.33333Z"
              fill="#7C8291"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
